src/transformStateWithClones.js

```diff
@@ -7,7 +7,35 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateVirsions = [];
+  const copyState = { ...state };
+
+  for (const action of actions) {
+    Object.assign(copyState, action.extraData);
+
+    switch (action.type) {
+      case 'addProperties':
+        stateVirsions.push(Object.assign({}, copyState));
+        break;
+
+      case 'removeProperties':
+        for (const property of action.keysToRemove) {
+          delete copyState[property];
+        };
+        stateVirsions.push(Object.assign({}, copyState));
+        break;
+
+      case 'clear':
+        Object.keys(copyState).forEach(key => delete copyState[key]);
+        stateVirsions.push({});
+        break;
+
+      default:
+        break;
```

Comment:
You should throw an Error is something is wrong, use `throw new Error(error description)` for that